src/InMemoryEventStore.php

```diff
@@ -283,39 +286,53 @@ public function transactional(callable $callable)
         return $result ?: true;
     }
 
-    public function createQuery(): Query
+    public function createQuery(QueryFactory $factory = null): Query
     {
-        return new InMemoryEventStoreQuery($this);
+        if (null === $factory) {
+            $factory = $this->getDefaultQueryFactory();
+        }
+
+        return $factory->factory();
     }
 
-    public function createProjection(string $name, ProjectionOptions $options = null): Projection
-    {
-        if (null === $options) {
-            $options = new ProjectionOptions();
+    public function createProjection(
+        string $name,
+        ProjectionOptions $options = null,
+        ProjectionFactory $factory = null
+    ): Projection {
+        if (null === $factory) {
+            $factory = $this->getDefaultProjectionFactory();
         }
 
-        return new InMemoryEventStoreProjection(
-            $this,
-            $name,
-            $options->cacheSize(),
-            $options->sleep()
-        );
+        return $factory->factory($name, $options);
     }
 
-    public function createReadModelProjection(string $name, ReadModel $readModel, ProjectionOptions $options = null): ReadModelProjection
-    {
-        if (null === $options) {
-            $options = new ProjectionOptions();
+    public function createReadModelProjection(
+        string $name,
+        ReadModel $readModel,
+        ProjectionOptions $options = null,
+        ReadModelProjectionFactory $factory = null
+    ): ReadModelProjection {
+        if (null === $factory) {
+            $factory = $this->getDefaultReadModelProjectionFactory();
         }
 
-        return new InMemoryEventStoreReadModelProjection(
-            $this,
-            $name,
-            $readModel,
-            $options->cacheSize(),
-            $options->persistBlockSize(),
-            $options->sleep()
-        );
+        return $factory->factory($name, $readModel, $options);
+    }
+
+    public function getDefaultQueryFactory(): QueryFactory
+    {
+        return new InMemoryEventStoreQueryFactory($this);
+    }
+
```

Comment:
@prolic We don't need everytime a new instance (especially with invoke), should be cached. What do you think?